community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -0,0 +1,481 @@
+            // This index already exists, verify the header with what we got passed into the constructor this time
+
+            try
+            {
+                // Read header
+                long layoutIdentifier;
+                int majorVersion;
+                int minorVersion;
+                try ( PageCursor metaCursor = openMetaPageCursor( pagedFile ) )
+                {
+                    do
+                    {
+                        pageSize = metaCursor.getInt();
+                        rootId = metaCursor.getLong();
+                        lastId = metaCursor.getLong();
+                        layoutIdentifier = metaCursor.getLong();
+                        majorVersion = metaCursor.getInt();
+                        minorVersion = metaCursor.getInt();
+                        layout.readMetaData( metaCursor );
+                    }
+                    while ( metaCursor.shouldRetry() );
+                }
+                if ( layoutIdentifier != layout.identifier() )
+                {
+                    throw new IllegalArgumentException( "Tried to open " + indexFile + " using different layout "
+                            + layout.identifier() + " than what it was created with " + layoutIdentifier );
+                }
+                if ( majorVersion != layout.majorVersion() || minorVersion != layout.minorVersion() )
+                {
+                    throw new IllegalArgumentException( "Index is of another version than the layout " +
+                            "it tries to be opened with. Index version is [" + majorVersion + "." + minorVersion + "]" +
+                            ", but tried to load the index with version [" +
+                            layout.majorVersion() + "." + layout.minorVersion() + "]" );
+                }
+                // This index was created with another page size, re-open with that actual page size
+                if ( pageSize != pageCache.pageSize() )
+                {
+                    if ( pageSize > pageCache.pageSize() )
+                    {
+                        throw new IllegalStateException( "Index was created with page size:" + pageSize
+                                + ", but page cache used to open it this time has a smaller page size:"
+                                + pageCache.pageSize() + " so cannot be opened" );
+                    }
+                    pagedFile.close();
+                    pagedFile = pageCache.map( indexFile, pageSize );
+                }
+                return pagedFile;
+            }
+            catch ( Throwable t )
+            {
+                try
+                {
+                    pagedFile.close();
+                }
+                catch ( IOException e )
+                {
+                    t.addSuppressed( e );
+                }
+                throw t;
+            }
+        }
+        catch ( NoSuchFileException e )
```

Comment:
Letting the page cache throwing this is actually better because of potentially different hardware configuration the page cache may have. A file system instance has a more limited view of the world, compared to the page cache.